src/GraphQL/MemoryExtensions.cs

```diff
@@ -0,0 +1,53 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+
+namespace GraphQL
+{
+    /// <summary>
+    /// Provides extension methods for working with arrays and pools.
+    /// </summary>
+    public static class MemoryExtensions
+    {
+        private static readonly ConcurrentDictionary<Type, Action<Array>> _delegates = new ConcurrentDictionary<Type, Action<Array>>();
+        private static readonly Func<Type, Action<Array>> _factory = CreateDelegate;
+
+        internal static void Return(this Array array) => _delegates.GetOrAdd(array.GetType(), _factory)(array);
```

Comment:
```suggestion
        internal static void Return(this Array array) => _delegates.GetOrAdd(array.GetType(), CreateDelegate)(array);
```

Why create `_factory`? Maybe rename `CreateDelegate` to `CreateDelegateFactory` if you want.

Reply:
https://github.com/dotnet/roslyn/issues/39869

I also wrote this comment for BasicVisitor:
![изображение](https://user-images.githubusercontent.com/21261007/104382557-5b6b6e80-553f-11eb-844e-810c66502ec0.png)